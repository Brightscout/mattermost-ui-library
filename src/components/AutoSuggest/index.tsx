import React, {useCallback, useEffect, useState, useRef} from 'react';

// Constants
import {DefaultCharThresholdToShowSuggestions} from '@Constants';

export type AutoSuggestProps = {
    inputValue: string;
    onInputValueChange: (newValue: string) => void;
    onOptionClick: (suggestion: Record<string, string>) => void;
    placeholder?: string;
    suggestionConfig: {
        suggestions: Record<string, string>[],
        renderValue: (suggestion: Record<string, string>) => string | JSX.Element;
    };
    loadingSuggestions?: boolean;
    charThresholdToShowSuggestions?: number;
    disabled?: boolean;
    error?: boolean | string;
    required?: boolean;
    className?: string;
}

const AutoSuggest = ({
    inputValue,
    onInputValueChange,
    placeholder,
    suggestionConfig,
    loadingSuggestions = false,
    charThresholdToShowSuggestions = DefaultCharThresholdToShowSuggestions,
    disabled = false,
    error,
    required,
    className = '',
    onOptionClick,
}: AutoSuggestProps) => {
    const [showSuggestions, setShowSuggestions] = useState(false);
    const [focused, setFocused] = useState(false);
    const textInputFieldRef = useRef<HTMLInputElement>(null);
    const autoSuggestRef = useRef<HTMLDivElement>(null);

    const {suggestions, renderValue} = suggestionConfig;

    // Show suggestions depending on the input value, number of characters and whether the input is in focused state
    useEffect(() => {
        setShowSuggestions(inputValue.length >= charThresholdToShowSuggestions && focused);
    }, [charThresholdToShowSuggestions, focused, inputValue]);

    useEffect(() => {
        if (focused) {
            textInputFieldRef.current?.focus();
        } else {
            textInputFieldRef.current?.blur();
        }
    }, [focused]);

    // Close the auto-suggest popover when the user clicks outside
    useEffect(() => {
        const handleCloseAutoSuggest = (e: MouseEvent) => !autoSuggestRef.current?.contains(e.target as Element) && setFocused(false);

        document.addEventListener('click', handleCloseAutoSuggest);

        return () => document.removeEventListener('click', handleCloseAutoSuggest);
    }, []);

    const handleSuggestionClick = useCallback((suggestedValue: Record<string, string>) => {
        onOptionClick(suggestedValue);
        setFocused(false);
    }, [onOptionClick]);

    // Prevent the text input field(which is the field visible in the UI) from blurring if "focused" is set to "true"
    const handleBlur = useCallback(() => focused && textInputFieldRef.current?.focus(), [focused]);

    return (
        <div
            className={
                `auto-suggest position-relative margin-bottom-25
                ${disabled && 'auto-suggest--disabled cursor-not-allowed'}
                ${error && 'auto-suggest--error'}
                ${className}
                `
            }
            ref={autoSuggestRef}
        >
            <div
                className={
                    `auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    ${focused && 'auto-suggest__field--focused'}
                    ${disabled && 'cursor-not-allowed'}
                    `
                }
            >
                <input
                    type='checkbox'
                    className='auto-suggest__toggle-input cursor-text'
                    checked={focused}
                    disabled={disabled}
                    onClick={() => setFocused(true)}
                    readOnly={true}
                />
                <input
                    ref={textInputFieldRef}
                    placeholder={`${placeholder ?? ''}${required ? '*' : ''}`}
                    value={inputValue}
                    onChange={(e) => onInputValueChange(e.target.value)}
                    onBlur={handleBlur}
                    className={`auto-suggest__input padding-0 ${disabled && 'cursor-not-allowed'}`}
                    disabled={disabled}
                />
                {loadingSuggestions ? (
                    <div className='auto-suggest__loader'/>
                ) : (
                    <i
                        className={
                            `fa fa-angle-down auto-suggest__field-angle font-20
                            ${showSuggestions && 'auto-suggest__field-angle--rotated'}
                            `
                        }
                    />
                )}
            </div>
            {inputValue.length < charThresholdToShowSuggestions && focused && (
                <p className='auto-suggest__get-suggestion-warn font-12'>
                    {`Please enter at least ${charThresholdToShowSuggestions} characters to get suggestions.`}
                </p>
            )}
            <ul className={`auto-suggest__suggestions padding-0 ${showSuggestions && 'auto-suggest__suggestions--open'}`}>
                {suggestions.map((suggestion) => (
                    <li
                        key={String(renderValue(suggestion))}
                        onClick={() => handleSuggestionClick(suggestion)}
                        className='auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0'
                    >
                        {renderValue(suggestion)}
                    </li>
                ))}
                {!suggestions.length && <li className='auto-suggest__suggestion cursor-pointer padding-v-10 padding-h-25 margin-0'>{'Nothing to show'}</li>}
            </ul>
            {typeof error === 'string' && <p className='auto-suggest__err-text error-text margin-top-5 font-14'>{error}</p>}
        </div>
    );
};

export default AutoSuggest;
