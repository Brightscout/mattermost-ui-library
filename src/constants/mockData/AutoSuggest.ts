// Mock data for "AutoSuggest" component tests
export const onInputValueChange = jest.fn();
export const onOptionClick = jest.fn();
export const suggestionConfig = {
    suggestions: [{
        label: 'Label 1',
        value: 'Value 1',
    }, {
        label: 'Label 2',
        value: 'Value 2',
    }, {
        label: 'Label 3',
        value: 'Value 3',
    }],
    renderValue: (suggestion: Record<string, string>) => suggestion.label,
};
