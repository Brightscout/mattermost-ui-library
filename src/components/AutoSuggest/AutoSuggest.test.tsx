import React from 'react';
import {mount, ReactWrapper} from 'enzyme';

import AutoSuggest, {AutoSuggestProps} from '.';

// Mock functions
const onInputValueChange = jest.fn();
const onOptionClick = jest.fn();

// Mock data
const suggestionConfig = {
    suggestions: [{
        label: 'Label 1',
        value: 'Value 1',
    }, {
        label: 'Label 2',
        value: 'Value 2',
    }, {
        label: 'Label 3',
        value: 'Value 3',
    }],
    renderValue: (suggestion: Record<string, string>) => suggestion.label,
};

type AutoSuggestComponentProps = Omit<AutoSuggestProps, 'inputValue' | 'onInputValueChange' | 'onOptionClick' | 'suggestionConfig'> & {
    emptySuggestions?: boolean
};

const AutoSuggestComponent = ({emptySuggestions, ...props}: AutoSuggestComponentProps) => (
    <AutoSuggest
        {...props}
        inputValue='Label'
        onInputValueChange={onInputValueChange}
        onOptionClick={onOptionClick}
        suggestionConfig={emptySuggestions ? {suggestions: [], renderValue: () => ''} : suggestionConfig}
    />
);

const clickAutoSuggest = (component: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>) => component.find('input[type="checkbox"]').simulate('click');

describe('AutoSuggest', () => {
    let component: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    beforeEach(() => {
        component = mount(<AutoSuggestComponent/>);
    });

    afterEach(() => {
        component.unmount();
    });

    it('Should render correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('Should show the loader when "loading" prop is passed', () => {
        const wrapper = mount(<AutoSuggestComponent loadingSuggestions={true}/>);
        expect(wrapper.find('.auto-suggest__loader').exists()).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('Should show the error message when "error" prop is passed', () => {
        const errorMsg = 'Error Message';
        const wrapper = mount(<AutoSuggestComponent error={errorMsg}/>);
        expect(wrapper.find('.auto-suggest__err-text').text().includes(errorMsg)).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('Should hide the suggestions when the input isn\'t focused', () => {
        expect(component.find('ul.auto-suggest__suggestions').hasClass('auto-suggest__suggestions--open')).toBeFalsy();
    });

    it('Should hide the character threshold warning when the input isn\'t focused', () => {
        const wrapper = mount(<AutoSuggestComponent charThresholdToShowSuggestions={8}/>);
        expect(wrapper.find('.auto-suggest__get-suggestion-warn').exists()).toBeFalsy();
        wrapper.unmount();
    });

    it('Should be in focused state when clicked', () => {
        clickAutoSuggest(component);
        const textField = component.find('input.auto-suggest__input');
        expect(textField.is(':focus')).toBe(true);
    });

    it('Should not show the suggestions when the input is focused but character threshold isn\'t reached', () => {
        const wrapper = mount(<AutoSuggestComponent charThresholdToShowSuggestions={8}/>);
        clickAutoSuggest(wrapper);
        expect(wrapper.find('ul.auto-suggest__suggestions').hasClass('auto-suggest__suggestions--open')).toBeFalsy();
        expect(wrapper.find('.auto-suggest__get-suggestion-warn').exists()).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('Should show the suggestions when the input is focused and character threshold is reached', () => {
        clickAutoSuggest(component);
        expect(component.find('ul.auto-suggest__suggestions').hasClass('auto-suggest__suggestions--open')).toBeTruthy();
        expect(component.find('.auto-suggest__suggestion')).toHaveLength(3);
        expect(component.find('.auto-suggest__get-suggestion-warn').exists()).toBeFalsy();
        expect(component).toMatchSnapshot();
    });

    it('Should show the empty state when "suggestions" are empty and character threshold is reached', () => {
        const wrapper = mount(<AutoSuggestComponent emptySuggestions={true}/>);
        clickAutoSuggest(wrapper);
        expect(wrapper.find('ul.auto-suggest__suggestions').hasClass('auto-suggest__suggestions--open')).toBeTruthy();
        expect(wrapper.find('.auto-suggest__suggestion').exists()).toBeTruthy();
        expect(wrapper.find('.auto-suggest__suggestion').text().includes('Nothing to show')).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });
});
