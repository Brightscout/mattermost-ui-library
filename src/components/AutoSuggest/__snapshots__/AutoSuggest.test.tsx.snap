// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoSuggest Should not show the suggestions when the input is focused but character threshold isn't reached 1`] = `
<AutoSuggestComponent
  charThresholdToShowSuggestions={8}
>
  <AutoSuggest
    charThresholdToShowSuggestions={8}
    inputValue="Label"
    onInputValueChange={[MockFunction]}
    onOptionClick={[MockFunction]}
    suggestionConfig={
      Object {
        "renderValue": [Function],
        "suggestions": Array [
          Object {
            "label": "Label 1",
            "value": "Value 1",
          },
          Object {
            "label": "Label 2",
            "value": "Value 2",
          },
          Object {
            "label": "Label 3",
            "value": "Value 3",
          },
        ],
      }
    }
  >
    <div
      className="auto-suggest position-relative margin-bottom-25
                false
                undefined
                
                "
    >
      <div
        className="auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    auto-suggest__field--focused
                    false
                    "
      >
        <input
          checked={true}
          className="auto-suggest__toggle-input cursor-text"
          disabled={false}
          onClick={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <input
          className="auto-suggest__input padding-0 false"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          value="Label"
        />
        <i
          className="fa fa-angle-down auto-suggest__field-angle font-20
                            false
                            "
        />
      </div>
      <p
        className="auto-suggest__get-suggestion-warn font-12"
      >
        Please enter at least 8 characters to get suggestions.
      </p>
      <ul
        className="auto-suggest__suggestions padding-0 false"
      >
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 1"
          onClick={[Function]}
        >
          Label 1
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 2"
          onClick={[Function]}
        >
          Label 2
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 3"
          onClick={[Function]}
        >
          Label 3
        </li>
      </ul>
    </div>
  </AutoSuggest>
</AutoSuggestComponent>
`;

exports[`AutoSuggest Should render correctly 1`] = `
<AutoSuggestComponent>
  <AutoSuggest
    inputValue="Label"
    onInputValueChange={[MockFunction]}
    onOptionClick={[MockFunction]}
    suggestionConfig={
      Object {
        "renderValue": [Function],
        "suggestions": Array [
          Object {
            "label": "Label 1",
            "value": "Value 1",
          },
          Object {
            "label": "Label 2",
            "value": "Value 2",
          },
          Object {
            "label": "Label 3",
            "value": "Value 3",
          },
        ],
      }
    }
  >
    <div
      className="auto-suggest position-relative margin-bottom-25
                false
                undefined
                
                "
    >
      <div
        className="auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    false
                    false
                    "
      >
        <input
          checked={false}
          className="auto-suggest__toggle-input cursor-text"
          disabled={false}
          onClick={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <input
          className="auto-suggest__input padding-0 false"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          value="Label"
        />
        <i
          className="fa fa-angle-down auto-suggest__field-angle font-20
                            false
                            "
        />
      </div>
      <ul
        className="auto-suggest__suggestions padding-0 false"
      >
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 1"
          onClick={[Function]}
        >
          Label 1
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 2"
          onClick={[Function]}
        >
          Label 2
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 3"
          onClick={[Function]}
        >
          Label 3
        </li>
      </ul>
    </div>
  </AutoSuggest>
</AutoSuggestComponent>
`;

exports[`AutoSuggest Should show the empty state when "suggestions" are empty and character threshold is reached 1`] = `
<AutoSuggestComponent
  emptySuggestions={true}
>
  <AutoSuggest
    inputValue="Label"
    onInputValueChange={[MockFunction]}
    onOptionClick={[MockFunction]}
    suggestionConfig={
      Object {
        "renderValue": [Function],
        "suggestions": Array [],
      }
    }
  >
    <div
      className="auto-suggest position-relative margin-bottom-25
                false
                undefined
                
                "
    >
      <div
        className="auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    auto-suggest__field--focused
                    false
                    "
      >
        <input
          checked={true}
          className="auto-suggest__toggle-input cursor-text"
          disabled={false}
          onClick={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <input
          className="auto-suggest__input padding-0 false"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          value="Label"
        />
        <i
          className="fa fa-angle-down auto-suggest__field-angle font-20
                            auto-suggest__field-angle--rotated
                            "
        />
      </div>
      <ul
        className="auto-suggest__suggestions padding-0 auto-suggest__suggestions--open"
      >
        <li
          className="auto-suggest__suggestion padding-v-10 padding-h-25 margin-0"
        >
          Nothing to show
        </li>
      </ul>
    </div>
  </AutoSuggest>
</AutoSuggestComponent>
`;

exports[`AutoSuggest Should show the error message when "error" prop is passed 1`] = `
<AutoSuggestComponent
  error="Error Message"
>
  <AutoSuggest
    error="Error Message"
    inputValue="Label"
    onInputValueChange={[MockFunction]}
    onOptionClick={[MockFunction]}
    suggestionConfig={
      Object {
        "renderValue": [Function],
        "suggestions": Array [
          Object {
            "label": "Label 1",
            "value": "Value 1",
          },
          Object {
            "label": "Label 2",
            "value": "Value 2",
          },
          Object {
            "label": "Label 3",
            "value": "Value 3",
          },
        ],
      }
    }
  >
    <div
      className="auto-suggest position-relative margin-bottom-25
                false
                auto-suggest--error
                
                "
    >
      <div
        className="auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    false
                    false
                    "
      >
        <input
          checked={false}
          className="auto-suggest__toggle-input cursor-text"
          disabled={false}
          onClick={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <input
          className="auto-suggest__input padding-0 false"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          value="Label"
        />
        <i
          className="fa fa-angle-down auto-suggest__field-angle font-20
                            false
                            "
        />
      </div>
      <ul
        className="auto-suggest__suggestions padding-0 false"
      >
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 1"
          onClick={[Function]}
        >
          Label 1
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 2"
          onClick={[Function]}
        >
          Label 2
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 3"
          onClick={[Function]}
        >
          Label 3
        </li>
      </ul>
      <p
        className="auto-suggest__err-text error-text margin-top-5 font-14"
      >
        Error Message
      </p>
    </div>
  </AutoSuggest>
</AutoSuggestComponent>
`;

exports[`AutoSuggest Should show the loader when "loading" prop is passed 1`] = `
<AutoSuggestComponent
  loadingSuggestions={true}
>
  <AutoSuggest
    inputValue="Label"
    loadingSuggestions={true}
    onInputValueChange={[MockFunction]}
    onOptionClick={[MockFunction]}
    suggestionConfig={
      Object {
        "renderValue": [Function],
        "suggestions": Array [
          Object {
            "label": "Label 1",
            "value": "Value 1",
          },
          Object {
            "label": "Label 2",
            "value": "Value 2",
          },
          Object {
            "label": "Label 3",
            "value": "Value 3",
          },
        ],
      }
    }
  >
    <div
      className="auto-suggest position-relative margin-bottom-25
                false
                undefined
                
                "
    >
      <div
        className="auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    false
                    false
                    "
      >
        <input
          checked={false}
          className="auto-suggest__toggle-input cursor-text"
          disabled={false}
          onClick={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <input
          className="auto-suggest__input padding-0 false"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          value="Label"
        />
        <div
          className="auto-suggest__loader"
        />
      </div>
      <ul
        className="auto-suggest__suggestions padding-0 false"
      >
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 1"
          onClick={[Function]}
        >
          Label 1
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 2"
          onClick={[Function]}
        >
          Label 2
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 3"
          onClick={[Function]}
        >
          Label 3
        </li>
      </ul>
    </div>
  </AutoSuggest>
</AutoSuggestComponent>
`;

exports[`AutoSuggest Should show the suggestions when the input is focused and character threshold is reached 1`] = `
<AutoSuggestComponent>
  <AutoSuggest
    inputValue="Label"
    onInputValueChange={[MockFunction]}
    onOptionClick={[MockFunction]}
    suggestionConfig={
      Object {
        "renderValue": [Function],
        "suggestions": Array [
          Object {
            "label": "Label 1",
            "value": "Value 1",
          },
          Object {
            "label": "Label 2",
            "value": "Value 2",
          },
          Object {
            "label": "Label 3",
            "value": "Value 3",
          },
        ],
      }
    }
  >
    <div
      className="auto-suggest position-relative margin-bottom-25
                false
                undefined
                
                "
    >
      <div
        className="auto-suggest__field padding-h-20 cursor-pointer d-flex align-items-center justify-content-between
                    auto-suggest__field--focused
                    false
                    "
      >
        <input
          checked={true}
          className="auto-suggest__toggle-input cursor-text"
          disabled={false}
          onClick={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <input
          className="auto-suggest__input padding-0 false"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          value="Label"
        />
        <i
          className="fa fa-angle-down auto-suggest__field-angle font-20
                            auto-suggest__field-angle--rotated
                            "
        />
      </div>
      <ul
        className="auto-suggest__suggestions padding-0 auto-suggest__suggestions--open"
      >
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 1"
          onClick={[Function]}
        >
          Label 1
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 2"
          onClick={[Function]}
        >
          Label 2
        </li>
        <li
          className="auto-suggest__suggestion text-ellipsis cursor-pointer padding-v-10 padding-h-25 margin-0"
          key="Label 3"
          onClick={[Function]}
        >
          Label 3
        </li>
      </ul>
    </div>
  </AutoSuggest>
</AutoSuggestComponent>
`;
